### DISCUSSION TITLE
Is `offline` mode really offline?

### DISCUSSION THREAD

**Discussion by LorenzoBettini** [eclipse-tycho/tycho#3205] [Q&A]
I've just migrated to Tycho 4.0.4 and in my GitHub Actions workflow, when I get a cache hit, I set the Maven option `--offline`. I the log I can see

```
[INFO] ### Using TychoRepositoryTransport for remote P2 access ###
[INFO]     Cache location:         /home/runner/.m2/repository/.cache/tycho
[INFO]     Transport mode:         offline
[INFO]     Http Transport type:    Java11Client
[INFO]     Update mode:            forced
```

So it's being `offline`, but then I can see

```
[INFO] Downloading from p2: https://download.eclipse.org/releases/2023-12/p2.index
[INFO] Downloaded from p2: https://download.eclipse.org/releases/2023-12/p2.index (122 bytes at 0 bytes/s)
[INFO] Downloading from p2: https://download.eclipse.org/technology/epp/packages/2023-12/p2.index
...
```

So it looks like it's not really offline?
Am I missing something? Or is it the expected behavior?

thanks in advance

**Comment by laeubi** [comment]
The message is printed independently, offline only means that the `CacheManager` will always use a cached file and never try to contact a remote. In online mode it will depending on the cache state maybe contact a remote if data is not already in the cache, outdated or update is forced.

**Comment by LorenzoBettini** [comment]
So, why does it contact update sites for `p2.index` (I see there's `forced`). Is there a way to have a pure offline mode?

**Comment by laeubi** [comment]
> So, why does it contact update sites for `p2.index` (I see there's `forced`). Is there a way to have a pure offline mode?

What do you mean by "forced"? All files (including `p2.index`) are cached (or could be at least) so it heavily depends on where you see the message and how you determine that it is not cached already.

**Comment by LorenzoBettini** [comment]
If you see above it says "update mode: forced".

In general, I was asking if there's a way to have a full offline mode, where no remote site is contacted (like maven offline mode).

**Comment by laeubi** [comment]
> If you see above it says "update mode: forced".

That's a different topic, and cause because you passing `-U` to maven what forces Tycho (+Maven) to not rely on the cached state but ask the server for updates what is a bit contradicting to be combined with `--offline` (both in maven and Tycho).

Regardless of that if you see "Downloading from p2" this does not imply the server is really contacted physically.

**Comment by LorenzoBettini** [comment]
You're right about `-U`: I though I removed them all but a few of them remained. I now removed them all and I can see `Update mode: cache first`. Sorry about that!

I still see downloading the `p2.index`. I can also see

```
[INFO] Downloading from p2: https://download.eclipse.org/releases/2023-12/p2.index
[INFO] Downloaded from p2: https://download.eclipse.org/releases/2023-12/p2.index (122 bytes at 0 bytes/s)
...
[INFO] Downloading from p2: https://download.eclipse.org/tools/orbit/simrel/orbit-aggregation/release/4.30.0/p2.index
[INFO] Downloaded from p2: https://download.eclipse.org/tools/orbit/simrel/orbit-aggregation/release/4.30.0/p2.index (172 bytes at 167 KB/s)
```

So, unless the INFO message is misleading it looks like the server is really contacted physically. Or can I assume it is not contacted?

**Comment by laeubi** [comment]
> I still see downloading the p2.index. I can also see

How do you "see" downloads?

> So, unless the INFO message is misleading it looks like the server is really contacted physically. Or can I assume it is not contacted?

As said if you specify `--offline` it will fail when the item is not from cache.

**Comment by laeubi** [comment]
By the way if you think message should be suppressed in offline we can certainly do that, just propose a PR for that.

**Comment by LorenzoBettini** [comment]
> > I still see downloading the p2.index. I can also see
> 
> How do you "see" downloads?

See the end of the second line:

```
[INFO] Downloading from p2: https://download.eclipse.org/tools/orbit/simrel/orbit-aggregation/release/4.30.0/p2.index
[INFO] Downloaded from p2: https://download.eclipse.org/tools/orbit/simrel/orbit-aggregation/release/4.30.0/p2.index (172 bytes at 167 KB/s)
```

The `(172 bytes at 167 KB/s)` looks like it downloaded something...

> 
> > So, unless the INFO message is misleading it looks like the server is really contacted physically. Or can I assume it is not contacted?
> 
> As said if you specify `--offline` it will fail when the item is not from cache.

I was thinking of Maven behavior when offline: of course it fails as you said, but it does not even try to establish a network connection. The INFO message above seems to imply a network connection.

**Comment by laeubi** [comment]
Best is to try locally with real offline and pass `-e` then one an see the offending part much easier than to guess from  log lines.

**Comment by LorenzoBettini** [comment]
@laeubi OK, I'll investigate further. However, in the meantime, I noted that in other projects (with Tycho 4.0.4) I don't see any logging about p2 bundles being downloaded (in the project where I noted the original issue instead I see the logging); is there any specific configuration to enable logging? I'm quite sure in the project where I see logging I haven't configured anything specific...
